Add Backend JWT Generation Documentation (#10)
wso2/docs-mg: en/docs/how-tos/generating-backend-jwt.md

# Generating Backend JWT 

You can use JWT Generation in API Microgateway to send a customized JWT to the backend with user preferred claims. You have three possible options when selecting a method to send a JWT to the backend.

- [Passing a token to the backend](#passing-a-token-to-the-backend)

> - [Generating a JWT inside Microgateway](#generating-a-jwt-inside-microgateway)

Author: I'll update and paraphrase it.

    "iss": "https://localhost:9443/oauth2/token",
    "exp": 1595451542,
    "iat": 1595447942,
    "jti": "0987c6d3-f11a-402c-8eb2-b21fb7b97f4d",
    "backendJwt": "eyJ4NXQiOiJNell4TW1Ga09HWXdNV0kwWldObU5EY3hOR1l3WW1NNFpUQTNNV0kyTkRBelpHUXpOR00wWkdSbE5qSmtPREZrWkRSaU9URmtNV0ZoTXpVMlpHVmxOZyIsImtpZCI6Ik16WXhNbUZrT0dZd01XSTBaV05tTkRjeE5HWXdZbU00WlRBM01XSTJOREF6WkdRek5HTTBaR1JsTmpKa09ERmtaRFJpT1RGa01XRmhNelUyWkdWbE5nX1JTMjU2IiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiJhZG1pbkBjYXJib24uc3VwZXIiLCJhdWQiOiJKekNMdUpxNmtISXBNaEs3QzVPaGlXTjdjUHNhIiwibmJmIjoxNTk1NDQ3OTQyLCJhenAiOiJKekNMdUpxNmtISXBNaEs3QzVPaGlXTjdjUHNhIiwic2NvcGUiOiJhbV9hcHBsaWNhdGlvbl9zY29wZSBkZWZhdWx0IiwiaXNzIjoiaHR0cHM6XC9cL2xvY2FsaG9zdDo5NDQzXC9vYXV0aDJcL3Rva2VuIiwiZXhwIjoxNTk1NDUxNTQyLCJpYXQiOjE1OTU0NDc5NDIsImp0aSI6IjA5ODdjNmQzLWYxMWEtNDAyYy04ZWIyLWIyMWZiN2I5N2Y0ZCJ9.Ewz50wbRkFnod4-l5iQzuayHG6O5820Ykg8fQcj_7mIBDAOnWCk7j2a9KOFlJtMyqUXfkchgZEOM0sN9o7GmyYPxpw7x0pjHalNlrdNX8uwnrwaRfgn7G_9d57iFEUTyo7RHfR8uYNJYzSjHgd4x-Vy0-8fZm2OErtoVA0gH7pwnhnE8x3AQOTnBKbGEsWRoR6UmRNGCDwr23GlQ5v7GbPj4puJiE5Z-p5DNOtWuWX1BOpn9IpNn3nio-9wvE9DQUssDojn2G6t8BlM9vv9gMJJT3zK64tfCoWuhTMPn74vs3zGuYy2KO7x7dtULcv2o56slFCqHAID44KuB93vS7A"
}
```

> After altering the configuration, you can run API Microgateway. When you are invoking a resource with a JWT token, you will be able to obtain the generated backend JWT.

Reviewer: Altering which configuration?

Author: I will section to the bottom with some improvements, like a conclusion to the doc.

### Generating a JWT inside Microgateway

Microgateway defines an [AbstractMGWJWTGenerator abstract class](https://github.com/wso2/product-microgateway/blob/master/components/micro-gateway-jwt-generator/src/main/java/org/wso2/micro/gateway/jwt/generator/AbstractMGWJWTGenerator.java) to write a custom JWT generator class and developers can extend this abstract class to implement their logic to generate the JWT. There are two abstract methods named populateStandardClaims and populateCustomClaims where you can write your logic to populate claims in the JWT. Furthermore, you have the capability to write your own logic overriding any of the current methods in the abstract class. 

> #### Writing a JWT generator

Reviewer: This is customization for option 2 of generating backend JWT. Is that correct?

Author: Yes, this is a sub-topic under option 2

> # Generating Backend JWT 

Reviewer 2: Change title to "Passing Enduser Attributes to the Backend Using JWT"

You can use JWT Generation in API Microgateway to send a customized JWT to the backend with user preferred claims. You have 2 possible options when selecting a method to send a JWT to the backend.

> - [Passing the Backed token in the JWT](#passing-the-backed-token-in-the-jwt)

Reviewer 3: ```suggestion
- [Extract backed token from the jwt access token](#extract-backed-token-from-the-jwt-access-token)
```

> - [Generating a JWT in Microgateway](#generating-a-jwt-in-microgateway)

Reviewer 3: Switch this topic as the first one

Apart from writing a JWT generator, developers have the capability use the default [MGWJWTGeneratorImpl class](https://github.com/wso2/product-microgateway/blob/master/components/micro-gateway-jwt-generator/src/main/java/org/wso2/micro/gateway/jwt/generator/MGWJWTGeneratorImpl.java) without implementing a custom class for JWT generation and manipulate the token, only tweaking the properties of the token using [JWT generator configurations](#jwt-generator-configurations).

#### Adding a custom JWT generator to the project

Once the custom JWT generator is written, the project should be built and the output jar should be placed in the `<MGW-project>/lib` directory. If third-party libraries are used when writing the JWT generator, these custom jars should also be placed in the same directory.

You can provide the classpath of the custom JWT generator in the **generatorImpl** configuration in the `<MICRO-GW-RUNTIME_HOME>/conf/micro-gw.conf` file. 
    

> ```yml

Reviewer 3: ```suggestion
```toml
```

[jwtGeneratorConfig]
  jwtGeneratorEnabled=true
  generatorImpl="sample.jwt.generator.SampleJWTGenerator"
```
Since, there are getters and setters for all the configuration properties in the abstract class you can use any of the configuration values to implement your logic in the custom JWT generator.

After enabling JWT generation from the configuration setting `jwtGeneratorEnabled` to true, start the API Microgateway runtime. When invoking a resource with a JWT token, you will be able to obtain the generated backend JWT.

### JWT Generator Configurations

You have the ability customize the token generation from the Microgateway runtime configuration in `<MICRO-GW-RUNTIME_HOME>/conf/micro-gw.conf`.

> ```yml

Reviewer 3: ```suggestion
```toml
```

  privateKeyAlias="ballerina"
  # JWT token expiry time - ms (valid only if the jwt generator caching mechanism is disabled)
  tokenExpiry=900000
  # Restricted claims as an array that should not be included in the backend JWT token
  # Example: restrictedClaims=["claim1","claim2","claim3"]
  restrictedClaims=[]
  # Token issuer standard claim
  issuer="wso2.org/products/am"
  # Token audience standard claim
  audience=["http://org.wso2.apimgt/gateway"]
  # JWT token generator implementation
  generatorImpl="org.wso2.micro.gateway.jwt.generator.MGWJWTGeneratorImpl"
  # JWT Generator cache configurations
  [jwtGeneratorConfig.jwtGeneratorCaching]
    # Enable jwt generator token caching
    tokenCacheEnable=true
    # Token cache expiry time (ms)
    tokenCacheExpiryTime=900000
    # Token cache capacity
    tokenCacheCapacity=10000
    # Token cache eviction factor
    tokenCacheEvictionFactor=0.25
```

Basically, outer `[jwtGeneratorConfig]` configuration contains the properties and the customizations of the generated JWT token whereas the inner `[jwtGeneratorConfig.jwtGeneratorCaching]` configuration handles the caching aspects of the generated JWT tokens.

> | Configuration | Description                         | Default |

Reviewer 3: Change to html table

| audience | Audience array of the generated JWT. ('aud' standard claim) | ["http://org.wso2.apimgt/gateway"] | 
| generatorImpl | JWT Generator implementation. (By default a JWT generator implementation is provided in Microgateway whereas you can add your custom implementation here) | org.wso2.micro.gateway.jwt.generator.MGWJWTGeneratorImpl | 
| tokenCacheEnable | Enable JWT Generator caching which enables to cache generated JWT so that it is not generating separate JWTs for each invocation. | true | 
| tokenCacheExpiryTime | Expiry time of the JWT Generator cache | 900000 | 
| tokenCacheCapacity | Capacity of the JWT Generator cache | 10000 | 
| tokenCacheEvictionFactor | Eviction factor of the JWT Generator cache | 0.25 | 

#### Setting the header of the backend request

When you are passing a token to the backend or when you are generating a JWT inside Microgateway, you can configure the header in which the JWT token will be passed to the backend. The following configuration should be added to the Microgateway runtime configuration in `micro-gw.conf`. (By default the header is set to X-JWT-Assertion)

> ``` yml

Reviewer 3: ```suggestion
```toml
```

| tokenCacheEnable | Enable JWT Generator caching which enables to cache generated JWT so that it is not generating separate JWTs for each invocation. | true | 
| tokenCacheExpiryTime | Expiry time of the JWT Generator cache | 900000 | 
| tokenCacheCapacity | Capacity of the JWT Generator cache | 10000 | 
| tokenCacheEvictionFactor | Eviction factor of the JWT Generator cache | 0.25 | 

After altering the configuration, you can run API Microgateway. When you are invoking a resource with a JWT token, you will be able to obtain the generated backend JWT.

### Generating a JWT using a custom JWT Generator

Microgateway defines an AbstractMGWJWTGenerator abstract class to write a custom JWT generator class. The abstract class is defined as below and the developers can extend this abstract class to implement their logic to generate the JWT. There are two abstract methods named populateStandardClaims and populateCustomClaims where you can write your logic to populate claims in the JWT. Furthermore, you have the capability to write your own logic overriding any of the current methods in the abstract class. 

> #### AbstractMGWJWTGenerator class

Reviewer: Point this to the mgw source.

     */
    public String encode(byte[] stringToBeEncoded) {
        return java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(stringToBeEncoded);
    }

    /**
     * Helper method to hexify a byte array.
     */
    public String hexify(byte bytes[]) {
        char[] hexDigits = {'0', '1', '2', '3', '4', '5', '6', '7',
                '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

        StringBuilder buf = new StringBuilder(bytes.length * 2);
        for (byte aByte : bytes) {
            buf.append(hexDigits[(aByte & 0xf0) >> 4]);
            buf.append(hexDigits[aByte & 0x0f]);
        }
        return buf.toString();
    }

    /**
     * Method to convert Java array to JSONArray.
     */
    public JSONArray arrayToJSONArray(Object[] objectArray) {
        JSONArray jsonArray = new JSONArray();
        jsonArray.addAll(Arrays.asList(objectArray));
        return jsonArray;
    }

    public abstract Map<String, Object> populateStandardClaims(Map<String, Object> jwtInfo);
    public abstract Map<String, Object> populateCustomClaims(Map<String, Object> jwtInfo,
                                                             ArrayList<String> restrictedClaims);
}
```

> #### Writing a JWT generator

Reviewer: Move this to samples and provide the link.
